Realm integration

Brahmancer opened the pull request:
Initial Pull request. Feel free to tear this code a new one :smile:

Commit:
Initial Branch Commit for Realm Integration.

Commit:
Naming and other small changes.

Commit:
More naming fixes

Commit:
Merge branch 'master' of https://github.com/ScienceSquad/Health into RealmIntergration

Commit:
Vague idea of how a Realm would be set up.

Commit:
Had a better idea of how realm would be set up using a Module constructor.
Also added tags because it's the right thing to do.

Commit:
Good idea how a module's Realm will be set up and used.
Next commits will used for testing! :D

Commit:
Fixing a Tag

Commit:
Well, at least it runs.

Commit:
Able to add mass values as well as update values.

Commit:
Cleaned up comments

Commit:
Abstraction is good.

Commit:
Abstraction is good 2: Electric Bugaloo.

Commit:
Initial RealmBackupAgent commit

Commit:
Adding closing functionality.

Commit:
Cleaning it up.

Commit:
Adding comments and init Realm Events ready to test.

Commit:
Realized I was being redundant with events.

Commit:
Things are looking like they're almost done.

Brahmancer commented:
@Wall-Dough @MrJohnson @colinrford @brandonschabell should also look and see if anything seems off.

Commit:
Comment changes. Nothing big.

avaidyam commented on Health/src/main/java/com/sciencesquad/health/health/Nutrition/RealmNutritionModel.java:
```diff
@@ -0,0 +1,57 @@
+ * Realm Model for Nutrition database.
+ * Things that are relevant to store for the Nutrition Module go here.
+ */
+public class RealmNutritionModel extends RealmObject {
```
Use annotations: `@Value.Immutable` to [mark this as a special immutable data object](http://immutables.github.io/).

Brahmancer replied:
So if we want to add the ability to have the user change their data for whatever reason (erring being human), would we still be able to do that with Immutable?

avaidyam replied:
Yes! Simple:

```
RealmNutritionModel model = getDataFromSomewhere();
RealmNutritionModel update = RealmNutritionModel.copy(model).calorieIntake(38574).build();
putDataSomewhere(update);
```

**Copying is always preferred over mutability.** ^remember that, kids.

avaidyam replied:
Turns out, Realm hates this and won't work with it. :cry:

avaidyam commented on Health/src/main/java/com/sciencesquad/health/health/Nutrition/RealmNutritionModel.java:
```diff
@@ -0,0 +1,57 @@
+ */
+public class RealmNutritionModel extends RealmObject {
+
+    private static final String TAG = "Realm Nutrition Model";
```
Model classes won't EVER interact directly to the user or developer, so there's no need for a tag.

avaidyam commented on Health/src/main/java/com/sciencesquad/health/health/Nutrition/RealmNutritionModel.java:
```diff
@@ -0,0 +1,57 @@
+    @NonNull
+    private Date date;
+
+    public RealmNutritionModel(){
```
Use the builder pattern. This comes for free when you use `@Value.Immutable`.

avaidyam commented on Health/src/main/java/com/sciencesquad/health/health/Nutrition/RealmNutritionModel.java:
```diff
@@ -0,0 +1,57 @@
+
+    private static final String TAG = "Realm Nutrition Model";
+
+    /**
```
Comments should go above EVERY field.

avaidyam commented on Health/src/main/java/com/sciencesquad/health/health/Nutrition/RealmNutritionModel.java:
```diff
@@ -0,0 +1,57 @@
+        this.date = rightNow.getTime();
+    }
+
+    public int getCalorieIntake(){ return calorieIntake; }
```
For any Model class, only `get...` methods are desired, since you never want to be modifying data without realizing it. Also, use `@Value.Immutable` which generates everything for you.

avaidyam commented on Health/src/main/java/com/sciencesquad/health/health/database/events/RealmUpdate.java:
```diff
@@ -0,0 +1,15 @@
+ * Event for updating a Realm in any abstract way.
+ */
+@Event.EventType @Value.Immutable
+public interface RealmUpdate extends Event {
```
Move all these events into your `RealmDataContext` class. They'll be auto-generated anyway.

Brahmancer replied:
What do you mean by this? Add the declarations of the events inside the RealmDataContext Class?

avaidyam replied:
Yup, just cut these declarations, paste em just underneath:

``` java
public class Foo extends Bar implements Qux {
    // declarations here...
    // code here...
}
```

avaidyam commented on Health/src/main/java/com/sciencesquad/health/health/database/events/RealmEmpty.java:
```diff
@@ -0,0 +1,16 @@
+ */
+@Value.Immutable @Event.EventType
+public interface RealmEmpty extends Event {
+    String RealmName();
```
Lowercase the `r` here in `RealmName`

avaidyam commented on Health/src/main/java/com/sciencesquad/health/health/database/AbstractDataContext.java:
```diff
@@ -0,0 +1,21 @@
+ *  - Update the database
+ *  - Query the database
+ */
+public abstract class AbstractDataContext {
```
Does this really solve all the problems for any storage medium? How do I set or get values by key? What other DB operations can I perform?

avaidyam commented on Health/src/main/java/com/sciencesquad/health/health/database/AbstractDataContext.java:
```diff
@@ -0,0 +1,21 @@
+ */
+public abstract class AbstractDataContext {
+
+    private static final String TAG = "Abstract Database Context";
```
`TAG` should always be `CurrentClass.class.className()`.

avaidyam commented on Health/src/main/java/com/sciencesquad/health/health/Nutrition/RealmNutritionModule.java:
```diff
@@ -0,0 +1,172 @@
+    @Override
+    public void closeRealm() {
+
+        /**
```
Comments go above the method, not inside; you might be thinking of Python :wink:

avaidyam commented on Health/src/main/java/com/sciencesquad/health/health/Nutrition/RealmNutritionModule.java:
```diff
@@ -0,0 +1,172 @@
+ */
+
+@RealmModule(classes = {RealmNutritionModel.class})
+public class RealmNutritionModule extends RealmDataContext{
```
So does every Module need to replicate this class for their data access? It seems really tedious...

avaidyam commented on Health/src/main/java/com/sciencesquad/health/health/Nutrition/NutritionModule.java:
```diff
@@ -0,0 +1,81 @@
+
+    }
+
+    protected void testNutritionModule(){
```
We should think about adding an `androidTests` folder and dumping a ton of unit tests there right off the bat.

Brahmancer replied:
How do you go about fixing this?

avaidyam replied:
That's a really good question. [Lemme Google that for us.](http://bfy.tw/4Nax)

Brahmancer replied:
I'm having trouble understanding how to get the test folder set up. Either we can try and get that pull in now or get these changes in and add that in later.

avaidyam replied:
We can add tests later. :shipit:

avaidyam commented on Health/src/main/java/com/sciencesquad/health/MainActivity.java:
```diff
@@ -34,6 +43,13 @@ protected void onCreate(Bundle savedInstanceState) {
 		NavigationView navigationView = (NavigationView) findViewById(R.id.nav_view);
 		navigationView.setNavigationItemSelectedListener(this);
+
+		RealmConfiguration defaultConfig = new RealmConfiguration.Builder(this)
```
This is a TERRIBLE idea since what you're saying is "Whenever this Activity is created, reconfigure and create a new `NutritionModule`". Serious flaw: the `Activity` will be recreated whenever you leave it for a while and come back (i.e. multitasking drawer).

Brahmancer replied:
So what would be a better way to approach this such that we only do this once?

avaidyam replied:
Either use a static block:

``` java
public class Foo {
    static {
        // only called once ever
    }

    // code here...
}
```

Or wait for the Module class to do the initialization procedures and hook into `BaseApplication.AppCreateEvent`...

Also, bet you didn't know this, but you can also use a fallthrough initializer.

``` java
public class Foo {
    static {
        // only called once ever
    }
    { // notice there's no "static" mark
        // called after the initializer no matter what
    }

    // code here...
}
```

avaidyam commented:
@Brahmancer Mauled your code a little. Perhaps your ego too, while I was at it. :laughing: 

But good job on the initial implementation; most of the issues were with Android specifics and code duplication here and there! :smile_cat:

Brahmancer commented:
It comes with the job. And someone's gotta make sure I'm not killing our code this early on :wink:

avaidyam commented:
:+1: :shipit: 

![](https://qph.is.quoracdn.net/main-qimg-c8781a4bb1f17e330b50cb35f851da05?convert_to_webp=true)

Brahmancer commented:
I'm having issues understanding the following: 
- Making Immutable objects without causing everything to not compile
- Events not being recognized. 

I feel like this is probably better solved in person because I honestly am lost why this is happening. Prime example why 4pm coding session is needed imo but that's just me.

avaidyam commented:
No, it's not just you. This is the exact reason I suggested it!!

Anyways, we'll meet and go over it on Tuesday.

Are you okay with me pushing commits to your branch?

> ## 

## 

Aditya.

Brahmancer commented:
Sure. Just let me know when you've updated files.

Commit:
Moved things around in packages and fixed comments.

Commit:
Fixing things that can be fixed. Mainly packaging issues.

Commit:
Cleanup again.

Commit:
Little more cleanup -- refactored stuff.

Brahmancer commented:
There's this random folder that has a garbage character in the file path which is causing the travis build to fail. How does one remove it.

avaidyam commented:
@Brahmancer Bruh I was working on that :( Y u do dis??

Brahmancer commented:
This was completely unintentional! At least in the sense to wreck things up. Feel free to discard all changes in my last commit if it's a problem.

Brahmancer commented:
Unintentional i

Commit:
Revert "Fixing things that can be fixed. Mainly packaging issues."

This reverts commit 420ae33aeb3669b79ede2a7e7ad707b7504bc456.

avaidyam commented:
@Brahmancer Okay, now take a look and see what changes you need to make.

Commit:
Merge branch 'RealmIntergration' of /Users/danielmiller/CS307/Health with conflicts.

Commit:
Compile again pls.

Commit:
The light is at the end of the tunnel (maybe)

Commit:
Merge remote-tracking branch 'origin/RealmIntergration' into RealmIntergration

Commit:
nutrition refactor and modifying unit test

Commit:
Modifying the abstraction slightly.

Commit:
Making things abstract electric bugaloo

Commit:
Send more help pls

Commit:
DROOGS

Commit:
nutrition pls

Commit:
Fixed packages

Commit:
I FEEXED IT

colinrford commented:
I think this might be good _>,.,<_/

Commit:
Slight tweak.

Commit:
slight tweak

slight tweak related to dan’s slight tweak. just wanted to include
myself in there

colinrford commented:
Are we scrapping the TypeToken idea or keeping it? As of right now it looks unused

avaidyam commented:
We'll be scrapping it but I need to come up with a solution. Currently working on another branch; go look at that one!

Commit:
I messed it all up... but it's nicer now.

Commit:
Travis pls.

Commit:
LOTS of cleaning up. Now, DataContext is an AutoCloseable Collection. RealmContext implements (by delegation) all of its collection-related methods, while still exposing the actual Realm and the SQL-style query.

Commit:
Merge remote-tracking branch 'origin/RealmIntergration' into RealmIntergration

# Conflicts:
#	Health/src/main/java/com/sciencesquad/health/data/RealmContext.java

Commit:
Wrapping things up.

Commit:
Adding a data failure event.

Commit:
Added Failures to DataOperations (mostly)

Commit:
Tiny clean up

Commit:
Tiny clean up

avaidyam committed:
Merge branch 'master' into RealmIntergration

Commit:
Clean up, clean up, everybody do your share...
